server/mediaserver.go

```diff
@@ -41,36 +42,39 @@ var ErrRTMPPlay = errors.New("ErrRTMPPlay")
 var ErrRoundInit = errors.New("ErrRoundInit")
 var ErrStorage = errors.New("ErrStorage")
 var ErrDiscovery = errors.New("ErrDiscovery")
+var ErrUnknownStream = errors.New("ErrUnknownStream")
 
 const HLSWaitInterval = time.Second
 const HLSBufferCap = uint(43200) //12 hrs assuming 1s segment
 const HLSBufferWindow = uint(5)
+const StreamKeyBytes = 6
 
 const SegLen = 4 * time.Second
-const HLSUnsubWorkerFreq = time.Second * 5
 const BroadcastRetry = 15 * time.Second
 
-var HLSWaitTime = time.Second * 45
 var BroadcastPrice = big.NewInt(1)
 var BroadcastJobVideoProfiles = []ffmpeg.VideoProfile{ffmpeg.P240p30fps4x3, ffmpeg.P360p30fps16x9}
-var MinDepositSegmentCount = int64(75)     // 5 mins assuming 4s segments
-var MinJobBlocksRemaining = big.NewInt(40) // 10 mins assuming 15s blocks
+var MinDepositSegmentCount = int64(75) // 5 mins assuming 4s segments
 var LastHLSStreamID core.StreamID
 var LastManifestID core.ManifestID
 
+type rtmpConnection struct {
+	nonce  uint64
+	stream stream.RTMPVideoStream
+	pl     core.PlaylistManager
+	sess   *BroadcastSession
+}
+
 type LivepeerServer struct {
-	RTMPSegmenter   lpmscore.RTMPSegmenter
-	LPMS            *lpmscore.LPMS
-	LivepeerNode    *core.LivepeerNode
-	VideoNonce      map[string]uint64
-	VideoNonceLock  *sync.Mutex
-	HttpMux         *http.ServeMux
-	CurrentPlaylist core.PlaylistManager
+	RTMPSegmenter lpmscore.RTMPSegmenter
+	LPMS          *lpmscore.LPMS
+	LivepeerNode  *core.LivepeerNode
+	HttpMux       *http.ServeMux
 
 	ExposeCurrentManifest bool
 
-	rtmpStreams                map[core.StreamID]stream.RTMPVideoStream
-	broadcastRtmpToManifestMap map[string]string
+	rtmpConnections map[core.ManifestID]*rtmpConnection
```

Comment:
Hadn't known about sync.Map -- thanks for the tip!

Looking into it, a `sync.Map` would probably work here, since we don't really need to assure mutual exclusion outside reading/writing the map itself. However, the [documentation](https://golang.org/pkg/sync/#Map) seems to discourage its use:

> Most code should use a plain Go map instead, with separate locking or coordination, for better type safety and to make it easier to maintain other invariants along with the map content. ... The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys

In particular, we don't fit the first case of an "always-growing cache". We sort of fit the second, but we'd still have concurrent access to the same set of keys (eg, playback).

Do you still feel that the sync.Map would be appropriate? I'm fairly indifferent; it would certainly help reduce manual locking (and the potential for mistakes associated with that), but I'm unsure if the caveats should be a concern. In particular, the lack of type safety makes me twitch a bit.

Reply:
When you mention playback requiring concurrent access to the same set of keys are you referring to the access in `getHLSMasterPlaylistHandler()`?

If we have concurrent access to the same set of keys as a result of the above, then I think it makes sense not to stick with a plain Go map + mutex. I originally suggested using `sync.Map` because I thought `rtmpConnections` fell under category 2 usage from the documentation. But if not, in the interest of keeping type safety and only using `sync.Map` when usage falls under one of those 2 categories I think sticking with a plain Go map + mutex makes sense.

Reply:
Yeah, there actually are a number of places that could read the same keys, although `getHLSMasterPlaylistHandler` and `getHLSMediaPlaylistHandler` are the major ones. Based on this conversation, I went ahead and changed those places to use a read lock (which is more correct in our case), here's the result (as diff'd from the streamid refactor branch):

```diff
diff --git a/server/mediaserver.go b/server/mediaserver.go
index 04a67632..155dd9ba 100644
--- a/server/mediaserver.go
+++ b/server/mediaserver.go
@@ -75,7 +75,7 @@ type LivepeerServer struct {
 	ExposeCurrentManifest bool
 
 	rtmpConnections map[core.ManifestID]*rtmpConnection
-	connectionLock  *sync.Mutex
+	connectionLock  *sync.RWMutex
 }
 
 func NewLivepeerServer(rtmpAddr string, httpAddr string, lpNode *core.LivepeerNode) *LivepeerServer {
@@ -91,7 +91,7 @@ func NewLivepeerServer(rtmpAddr string, httpAddr string, lpNode *core.LivepeerNo
 		opts.HttpMux = http.NewServeMux()
 	}
 	server := lpmscore.New(&opts)
-	return &LivepeerServer{RTMPSegmenter: server, LPMS: server, LivepeerNode: lpNode, HttpMux: opts.HttpMux, connectionLock: &sync.Mutex{}, rtmpConnections: make(map[core.ManifestID]*rtmpConnection)}
+	return &LivepeerServer{RTMPSegmenter: server, LPMS: server, LivepeerNode: lpNode, HttpMux: opts.HttpMux, connectionLock: &sync.RWMutex{}, rtmpConnections: make(map[core.ManifestID]*rtmpConnection)}
 }
 
 //StartServer starts the LPMS server
@@ -517,7 +517,7 @@ func getHLSMasterPlaylistHandler(s *LivepeerServer) func(url *url.URL) (*m3u8.Ma
 			manifestID = sid.ManifestID
 		}
 
-		s.connectionLock.Lock()
+		s.connectionLock.RLock()
 		defer s.connectionLock.Unlock()
 		cxn, ok := s.rtmpConnections[manifestID]
 		if !ok || cxn.pl == nil {
@@ -536,7 +536,7 @@ func getHLSMediaPlaylistHandler(s *LivepeerServer) func(url *url.URL) (*m3u8.Med
 	return func(url *url.URL) (*m3u8.MediaPlaylist, error) {
 		strmID := parseStreamID(url.Path)
 		mid := strmID.ManifestID
-		s.connectionLock.Lock()
+		s.connectionLock.RLock()
 		defer s.connectionLock.Unlock()
 		cxn, ok := s.rtmpConnections[mid]
 		if !ok || cxn.pl == nil {
@@ -589,7 +589,7 @@ func getHLSSegmentHandler(s *LivepeerServer) func(url *url.URL) ([]byte, error)
 func getRTMPStreamHandler(s *LivepeerServer) func(url *url.URL) (stream.RTMPVideoStream, error) {
 	return func(url *url.URL) (stream.RTMPVideoStream, error) {
 		mid := parseManifestID(url.Path)
-		s.connectionLock.Lock()
+		s.connectionLock.RLock()
 		cxn, ok := s.rtmpConnections[mid]
 		s.connectionLock.Unlock()
 		if !ok {
@@ -627,7 +627,7 @@ func (s *LivepeerServer) GetNodeStatus() *net.NodeStatus {
 	// not threadsafe; need to deep copy the playlist
 	m := make(map[string]*m3u8.MasterPlaylist, 0)
 
-	s.connectionLock.Lock()
+	s.connectionLock.RLock()
 	defer s.connectionLock.Unlock()
 	for _, cxn := range s.rtmpConnections {
 		if cxn.sess == nil || cxn.pl == nil {
@@ -641,7 +641,7 @@ func (s *LivepeerServer) GetNodeStatus() *net.NodeStatus {
 
 // Debug helpers
 func (s *LivepeerServer) LatestPlaylist() core.PlaylistManager {
-	s.connectionLock.Lock()
+	s.connectionLock.RLock()
 	defer s.connectionLock.Unlock()
 	cxn, ok := s.rtmpConnections[LastManifestID]
 	if !ok || cxn.pl == nil {

```